lib/charms/mongodb/v0/mongos.py

```diff
@@ -235,11 +216,37 @@ def _is_shard_draining(self, shard_name: str) -> bool:
             f"Shard {shard_name} not in cluster, could not retrieve draining status"
         )
 
-    def get_primary_shard(self) -> str:
-        """Processes sc_status and identifies the primary shard."""
-        # TODO Follow up PR, implement this function there is no MongoDB command to retrieve
-        # primary shard, this is possible with mongosh.
-        return False
+    def get_primary_shards(self) -> Optional[List[str]]:
+        """Returns the primary shards of the cluster.
+
+        In Sharded MongoDB clusters, mongos selects the primary shard when creating a new database
+        by picking the shard in the cluster that has the least amount of data. This means that:
+        1. There can be multiple primary shards in a cluster.
+        2. Until there is data written to the cluster there is effectively no primary shard.
+        """
+        databases_collection = self._get_databases_collection()
+        if databases_collection is None:
+            logger.info("No data written to sharded cluster yet, no primary shards.")
+            return None
+
+        return databases_collection.distinct("primary")
+
+    def get_databases_for_primary_shard(self, primary_shard) -> Optional[List[str]]:
+        """Returns a list of databases using the given shard as a primary shard."""
+        databases_collection = self._get_databases_collection()
+        if databases_collection is None:
+            return
+
+        return databases_collection.distinct("_id", {"primary": primary_shard})
+
+    def _get_databases_collection(self) -> collection.Collection:
+        """Returns the database collection if present."""
+        config_db = self.client["config"]
+        if "databases" not in config_db.list_collection_names():
```

Comment:
why not use `config_db.get_collection("databasess")`?

Reply:
because it is not guaranteed that the collection exists. The collection only gets created once a user writes data